### PR TITLE
Reimplement the dashed line effect

### PR DESCRIPTION
**Pull request by danielworkman**
* Could not find a way of getting a `Path` to correctly draw dashed effects so created a method to draw them manually
 * These are all drawn at once using the hardware accelerated `drawLines` so I do not expect any performance issues
 * Also fixed a bug whereby we were drawing splays with the dash effect as well

### CODE REVIEW

File: app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java [hunk 1]
```diff
@@ -952,8 +946,8 @@ private void drawCrossSections(
             drawLegs(canvas, projection, alpha);
 
             Coord2D viewStationLocation = surveyCoordsToViewCoords(surveyStationLocation);
-            drawLineAsPath(
-                    canvas, viewStationLocation, centreOnView, crossSectionConnectorPaint);
+            drawDashedLine(
+                    canvas, viewStationLocation, centreOnView, 5, crossSectionConnectorPaint);
```
Comment (richsmith) [inline app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java:950]: Magic number 5 should probably be a constant.

File: app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java [hunk 2]
```diff
@@ -1003,13 +997,11 @@ private void drawLegs(Canvas canvas, Space<Coord2D> space, int baseAlpha) {
                 paint = fade ? fadedLegPaint : legPaint;
             }
 
-            if (projectionType.isLegInPlane(leg)) {
-                canvas.drawLine((float)start.x, (float)start.y, (float)end.x, (float)end.y, paint);
+            if (leg.hasDestination() && !projectionType.isLegInPlane(leg)) {
```
Comment (richsmith) [inline app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java:1000]: Drawing splays that are out of the plane as dashed lines is supposed to happen (I think). Lets you see which measurements are shots up an aven or part of x-sections. You're usually drawing the walls in according to what's on your plane.

Comment (richsmith) [reply app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java:1000]: Something I've thought about but never bothered to play with to see if it feels right: have dotted lines for lines away from you (downwards on plan) and dashed lines coming towards you (upwards on plan).

File: app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java [hunk 3]
```diff
@@ -1394,26 +1386,39 @@ public void centreViewOnSurveyPoint(Coord2D point) {
...
-        paint.setAlpha(alpha);
-        canvas.drawLine(
-                (float)(start.x), (float)(start.y),
-                (float)(end.x), (float)(end.y),
-                paint);
-    }
+        double lineLength = Space2DUtils.getDistance(start, end);
+        int dashes = (int)(lineLength / dashLength);
 
+        Coord2D direction = end.minus(start).normalize();
```
Comment (richsmith) [inline app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java:1397]: Nice elegant way of doing it, but I'm a bit surprised to find Cartesian co ords being used as vectors. I guess they are technically vectors...

File: app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java [hunk 4]
```diff
@@ -27,10 +27,30 @@ public Coord2D minus(Coord2D coord2D) {
...
+
     public Coord2D scale(double scale) {
         return new Coord2D(x * scale, y * scale);
     }
 
+    public double mag() {
+        return Math.sqrt(this.x * this.x + this.y * this.y);
+    }
+
+    public Coord2D normalize() {
```
Comment (richsmith) [inline app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java:46]: SexyTopo is written in British English, old chap. One would normally write `normalise()`.

File: app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java [hunk 5]
```diff
@@ -27,10 +27,30 @@ public Coord2D minus(Coord2D coord2D) {
         return new Coord2D(this.x - coord2D.x, this.y - coord2D.y);
     }
 
+    public Coord2D multiply(Coord2D coord2D) {
+        return new Coord2D(this.x * coord2D.x, this.y * coord2D.y);
+    }
+
+    public Coord2D div(double divisor) {
```
Comment (richsmith) [inline app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java:34]: Do we need a `div` function? Could just `scale()` by `-mag`?

File: app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java (hunk 4 above)
Comment (danielworkman) [reply app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java:46]: Even though you've got other copy-paste Normalize() methods in here ;)

File: app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java (hunk 5 above)
Comment (danielworkman) [reply app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java:34]: Errrr that's not how maths works

Comment (richsmith) [reply app/src/main/java/org/hwyl/sexytopo/model/graph/Coord2D.java]: Oh yeah. You know what I meant :stuck_out_tongue:

**Review by richsmith — APPROVED** [review]
Good job :medal_sports:

**Pull request merged**